File: solutions/2-let-s-gossip/lib/gossip.js
const os = require('os');
const colors = require('colors');
const { updateState, renderState, sample } = require('./helpers');
const fs = require('fs');
const fetch = require('node-fetch');
const express = require('express');
const bodyParser = require('body-parser')
const app = express();
app.use(bodyParser.json());

// Set same PORT for everyone in Sinatra
const PORT = 1111;

// Get your local private IP
var ifaces = os.networkInterfaces();

const iface = Object.values(ifaces)
.map((faces) => faces.find(({family, internal}) => 'IPv4' === family && !internal))
.find((face) => face);

const HOST = iface ? iface.address : '127.0.0.1';

// Get the first peer from the command line
const PEER_HOST = process.argv[2] || HOST;

// Declare the State hash and set first value
const STATE = new Object();

const BANDS = fs.readFileSync("bands.txt", "utf8").split('\n').filter((band) => band !== '');
let favoriteBand = sample(BANDS);
let versionNumber = 0
console.log(`My favorite band, is ${favoriteBand.green}!`);

STATE[HOST] = [favoriteBand, versionNumber];
STATE[PEER_HOST] = null;

setInterval(() => {
  // TODO
  // Every 8 seconds change your own state to a new favorite band (don't forget to print something fun on the terminal)

  favoriteBand = sample(BANDS);
  versionNumber += 1;
  STATE[HOST] = [favoriteBand, versionNumber];
}, 8000);


setInterval(() => {
  const dupState = Object.assign({}, STATE);
  Object.keys(dupState).forEach((host) => {
    if (host === HOST) { return; }
    console.log(`Fetching update from ${host.green}`);
    fetch(`http://${host}:${PORT}/gossip`, { method: 'POST', body: JSON.stringify(STATE) })
    .then((res) => res.json())
    .then((newState) => {
      console.lgo(newState);
      updateState(newState, STATE);
    })
    .catch((err) => {
      delete STATE[host];
    });
  });
  renderState(STATE);
}, 3000);

app.post('/gossip', function (req, res) {
  // TODO get the JSON body, update the state with it and response the updated state in JSON
  const theirState = req.body;
  updateState(theirState, STATE)
  res.json(STATE);
})

app.listen(PORT);

File: solutions/2-let-s-gossip/lib/gossip.rb
require 'sinatra'
require 'faraday'
require 'socket'
require 'colorize'
require 'json'
require 'active_support/time'
require_relative 'helpers'

# Set same PORT for everyone in Sinatra
PORT = 1111
set :port, PORT
set :bind, '0.0.0.0'

# Get your local private IP
addr_infos = Socket.ip_address_list
HOST = addr_infos.find(&:ipv4_private?).ip_address

# Get the first peer from the command line
PEER_HOST = ARGV.first || HOST


# Declare the State hash and set first values
STATE = ThreadSafe::Hash.new
update_state(HOST => nil)
update_state(PEER_HOST => nil)

BANDS = File.readlines("bands.txt").map(&:chomp)
@favorite_band = BANDS.sample
@version_number = 0
puts "My favorite band, is #{@favorite_band.green}!"

update_state(HOST => [@favorite_band, @version_number])

every(8.seconds) do
  # COPY from the previous exercise
  @favorite_band = BANDS.sample
  @version_number += 1
  update_state(HOST => [@favorite_band, @version_number])

end

every(3.seconds) do
  # TODO Fetch all the hosts, except yourself and update the state.
  STATE.dup.keys.each do |host|
    next if host == HOST
    puts "Fetching update from #{host.green}"
    begin
      gossip_response = Faraday.post("http://#{host}:#{PORT}/gossip", JSON.dump(STATE), :content_type => 'application/json').body
      update_state(JSON.load(gossip_response))
    rescue Faraday::ConnectionFailed => e
      STATE.delete(host)
    end
  end
  render_state
end

# @param state
post '/gossip' do
  # TODO get the JSON body, update the state with it and response the updated state in JSON
  content_type :json
  their_state = JSON.parse(request.body.read)
  update_state(their_state)
  STATE.to_json
end
